File: utils/src/main/java/org/exthmui/utils/BaseEventListenersUtils.java
package org.exthmui.utils;

import androidx.annotation.NonNull;

import org.exthmui.utils.listeners.BaseEventListener;

import java.lang.reflect.InvocationTargetException;
import java.lang.reflect.Method;
import java.util.Collection;
import java.util.EventObject;

public abstract class BaseEventListenersUtils {
    public static boolean isThisListenerSuitable(@NonNull BaseEventListener listener, @NonNull Class<? extends BaseEventListener>[] listenerTypesAllowed) {
        for (Class<? extends BaseEventListener> t : listenerTypesAllowed) {
            if (t.isAssignableFrom(listener.getClass())) {
                return true;
            }
        }
        return false;
    }

    public static void notifyListeners(@NonNull EventObject event, @NonNull Collection<BaseEventListener> listeners) {
        for (BaseEventListener listener : listeners) {
            for (Class<? extends EventObject> t: listener.getEventToMethodMap().keySet()) {
                if (t.isAssignableFrom(event.getClass())){
                    Method[] methods = listener.getEventToMethodMap().get(t);
                    if (methods == null) break;
                    for (Method method: methods) {
                        try {
                            Method actualMethod = listener.getClass().getDeclaredMethod(method.getName(), method.getParameterTypes());

                            actualMethod.invoke(listener, event);
                        } catch (@NonNull NoSuchMethodException | IllegalAccessException e) {
                            e.printStackTrace();
                        } catch (InvocationTargetException e) {
                            Throwable targetException = e.getTargetException();
                            targetException.printStackTrace();
                        }
                    }
                }
            }
        }
    }
}

File: utils/src/main/java/org/exthmui/utils/BaseEventListenersUtils.kt
package org.exthmui.utils

import org.exthmui.utils.listeners.BaseEventListener
import java.lang.reflect.InvocationTargetException
import java.util.EventObject

object BaseEventListenersUtils {
    fun isThisListenerSuitable(
        listener: BaseEventListener,
        listenerTypesAllowed: Array<Class<out BaseEventListener>>
    ): Boolean {
        for (t in listenerTypesAllowed) {
            if (t.isAssignableFrom(listener.javaClass)) {
                return true
            }
        }
        return false
    }

    fun notifyListeners(event: EventObject, listeners: Collection<BaseEventListener>) {
        for (listener in listeners) {
            for (t in listener.eventToMethodMap.keys) {
                if (t.isAssignableFrom(event.javaClass)) {
                    val methods = listener.eventToMethodMap[t] ?: break
                    for (method in methods) {
                        try {
                            val actualMethod = listener.javaClass.getDeclaredMethod(
                                method.name,
                                *method.parameterTypes
                            )
                            actualMethod.invoke(listener, event)
                        } catch (e: NoSuchMethodException) {
                            e.printStackTrace()
                        } catch (e: IllegalAccessException) {
                            e.printStackTrace()
                        } catch (e: InvocationTargetException) {
                            val targetException = e.targetException
                            targetException.printStackTrace()
                        }
                    }
                }
            }
        }
    }
}
